### ISSUE TITLE
Single table delete cannot understand "orderBy" clause

### ISSUE THREAD

**Issue by niconoe-** [doctrine/doctrine2#11078]
### Feature Request

Regarding lots of DBMS, it is possible to define the order of deletion of rows during a DELETE statement on a single table with an ORDER BY clause. 

This ticket proposes to add the management of the ORDER BY clause on the DELETE statements that occur on single table only.

**NOTA:** PostgreSQL uses CURSOR to handle the order of deletion and its approach is completly different. I would suggest to ignore this feature for PostgreSQL if not already implemented and add a TODO to support it later.

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | no
| BC Break    | no

#### Summary

The problem is that when trying to add the ORDER BY clause on such DELETE statement via Doctrine/ORM, you have a syntax error: `[Syntax Error] line 0, col 78: Error: Expected end of string, got 'ORDER'`.  
This is not the only problem because it appears that the `Doctrine\ORM\Query\AST\DeleteStatement` class does not support the property `$orderByClause` also.  
Finally, when about to create the SQL to execute, the `Doctrine\ORM\Query\SqlWalker` class does not support the ORDER BY clause when walking on the DeleteStatement (method `walkDeleteStatement`).

#### Suggestion

1. Add the public property `$orderByClause` on the `Doctrine\ORM\Query\AST\DeleteStatement` class to support it:
```php
    /** @var OrderByClause|null */
    public $orderByClause;
```
2. Update `\Doctrine\ORM\Query\Parser::DeleteStatement` to support this new property and fill it properly:
```php
    public function DeleteStatement()
    {
        $deleteStatement = new AST\DeleteStatement($this->DeleteClause());

        $deleteStatement->whereClause = $this->lexer->isNextToken(Lexer::T_WHERE) ? $this->WhereClause() : null;
        $deleteStatement->orderByClause = $this->lexer->isNextToken(Lexer::T_ORDER) ? $this->OrderByClause() : null; // New line

        return $deleteStatement;
    }
```
3. Update `Doctrine\ORM\Query\SqlWalker::walkDeleteStatement` to make sure the `$orderByClause` property is handled when building the SQL:
```php
    public function walkDeleteStatement(AST\DeleteStatement $AST)
    {
        $this->useSqlTableAliases = false;
        $this->rsm->isSelect      = false;

        $sql = $this->walkDeleteClause($AST->deleteClause)
            . $this->walkWhereClause($AST->whereClause);
        if ($AST->orderByClause) {
            $sql .= $this->walkOrderByClause($AST->orderByClause);
        }
        return $sql;
    }
```

This was tested on my side with a DELETE on a single table, using MySQL and MariaDB, and worked like a charm. I did not perform any other tests with other DBMS, and I did not performed any tests with a DELETE on several tables (where the ORDER BY is not allowed).

Do you think it could be a new feature in order to get closer to what DBMS are proposing? Would you need any help (humbly) to provide a PR?

Thanks for reading and hoping this new feature will be a YES ! :slightly_smiling_face:

**Comment by YavorIvanov** [comment]
Just came across this the other day as well. I needed to provide order by in order to get the correct number of affecter rows due to a constraint deleting records and hence misreporting the affecter rows.

**Comment by niconoe-** [comment]
Hi again. No updates on this. I just wanted to notice that this issue of `Error: Expected end of string, got 'ORDER'` also occurs on `UPDATE` statements, which is allowed to give for some DBMS too:
- https://dev.mysql.com/doc/refman/9.0/en/update.html
- https://mariadb.com/docs/server/reference/sql-statements/data-manipulation/changing-deleting-data/update
- https://www.ibm.com/docs/en/db2/12.1.0?topic=statements-update
- …

Basically applying the same suggestion that I did before works (adapted for update, of course) on Mysql and MariaDb DBMS.

Thanks a lot!
